File: ios-app/CalendarioCiclismo/Models/Race.swift
import Foundation

/// Carrera profesional de ciclismo (tabla `races` en Supabase).
struct Race: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let nameEn: String?
    let abbrev: String?
    let uciCategory: String?
    let gender: String?
    let raceFormat: String?
    let countryCode: String?
    let colorHex: String?
    let logoUrl: String?
    let websiteUrl: String?
    let fcId: Int?
    let pcsSlug: String?
    let hideFlag: Bool
    let isGrandTour: Bool
    let isCancelled: Bool
    let startDate: String?   // YYYY-MM-DD
    let endDate: String?     // YYYY-MM-DD
    let year: Int?
    let slug: String?
    let originalName: String?
    let startlistImportedAt: String?
    let startlistProvisional: Bool?
    let enrichedStartlist: Bool?

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: Race, rhs: Race) -> Bool {
        lhs.id == rhs.id
    }

    // MARK: - Computed

    var isStageRace: Bool { raceFormat == "stage_race" }
    var isOneDay: Bool { raceFormat == "one_day" }
    var isFemale: Bool { gender == "female" }

    /// Nombre localizado según el idioma activo o el idioma del dispositivo.
    /// Muestra `nameEn` si el usuario tiene Premium+inglés O si el dispositivo
    /// está configurado en un idioma no español.
    var localizedName: String {
        if LocaleService.shouldShowEnglishContent, let en = nameEn, !en.isEmpty { return en }
        return name
    }

    /// Duración en días de la carrera.
    var durationDays: Int? {
        guard let s = startDate, let e = endDate,
              let start = DateFormatting.date(from: s),
              let end = DateFormatting.date(from: e) else { return nil }
        return Calendar.current.dateComponents([.day], from: start, to: end).day.map { $0 + 1 }
    }
}

File: android-app/app/src/main/java/app/calendariociclismo/android/data/model/Race.kt
package app.calendariociclismo.android.data.model

import app.calendariociclismo.android.util.DateFormatting
import app.calendariociclismo.android.util.LocaleHolder
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.concurrent.TimeUnit

/**
 * Carrera profesional de ciclismo (tabla `races` en Supabase).
 *
 * Port literal de [Race.swift][ios-app/CalendarioCiclismo/Models/Race.swift].
 */
@Serializable
data class Race(
    val id: String,
    val name: String,
    val nameEn: String? = null,
    val abbrev: String? = null,
    val uciCategory: String? = null,
    val gender: String? = null,
    val raceFormat: String? = null,
    val countryCode: String? = null,
    val colorHex: String? = null,
    val logoUrl: String? = null,
    val websiteUrl: String? = null,
    val fcId: Int? = null,
    val pcsSlug: String? = null,
    val hideFlag: Boolean = false,
    val isGrandTour: Boolean = false,
    val isNoClickable: Boolean = false,
    val isCancelled: Boolean = false,
    val startDate: String? = null, // YYYY-MM-DD
    val endDate: String? = null,   // YYYY-MM-DD
    val year: Int? = null,
    val slug: String? = null,
    val originalName: String? = null,
    val startlistImportedAt: String? = null,
    val startlistProvisional: Boolean = false,
    val enrichedStartlist: Boolean? = null,
    @SerialName("createdAt") val createdAt: String? = null,
) {
    val isStageRace: Boolean get() = raceFormat == "stage_race"
    val isOneDay: Boolean get() = raceFormat == "one_day"
    val isFemale: Boolean get() = gender == "female"

    /** Nombre localizado: muestra `nameEn` si la app está en inglés o si el
     *  dispositivo está configurado en un idioma no español. */
    val localizedName: String
        get() {
            if (LocaleHolder.shouldShowEnglishContent) {
                val en = nameEn
                if (!en.isNullOrEmpty()) return en
            }
            return name
        }

    /** Duración en días de la carrera (incluye ambos extremos). */
    val durationDays: Int?
        get() {
            val start = startDate?.let { DateFormatting.parseDateKey(it) } ?: return null
            val end = endDate?.let { DateFormatting.parseDateKey(it) } ?: return null
            val diff = TimeUnit.MILLISECONDS.toDays(end.time - start.time).toInt()
            return diff + 1
        }
}
